internal/server/recon/recognize.go

```diff
@@ -451,24 +461,78 @@ func getNumSpansForContainedIn(spans []*pb.TokenSpans_Span, startIdx int) int {
 	return 0
 }
 
-func combineContainedInSingle(
-	spans []*pb.TokenSpans_Span, startIdx, numSpans int) *pb.TokenSpans_Span {
+func combineTokenSpans(spans []*pb.TokenSpans_Span, startIdx, numSpans int) *pb.TokenSpans_Span {
 	startSpan := spans[startIdx]
-	endSpan := spans[startIdx+numSpans-1]
-
+	
 	res := &pb.TokenSpans_Span{Tokens: startSpan.Tokens}
 	for i := 1; i < numSpans; i++ {
 		res.Tokens = append(res.Tokens, spans[startIdx+i].GetTokens()...)
 	}
 
-	// This map is used to collect all the places for the combined span, with dedup.
+	return combineContainedInTokens(res, spans, startIdx, numSpans)
+}
+
+func getNextPlaceTokenSpan(spans []*pb.TokenSpans_Span, startIdx int) (*pb.TokenSpans_Span, int) {
+	for i := startIdx; i < len(spans); i++ {
+		if len(spans[i].GetPlaces()) > 0 {
+			return spans[i], i
+		}
+	}
+	return nil, -1
+}
+
+// Generates a map of dcid to the corresponding RecogPlace objects found, and a second map containing the nextContainingPlaces to the starting DCIDs.
+func findRecogPlaces(startSpan, nextSpan *pb.TokenSpans_Span) (map[string]*pb.RecogPlace, map[string][]string) {
 	dcidToRecogPlaces := map[string]*pb.RecogPlace{}
+	nextContainingPlaceToStartDcid :=map[string][]string{}
 
 	for _, p1 := range startSpan.GetPlaces() {
 		for _, containingPlace := range p1.GetContainingPlaces() {
-			for _, p2 := range endSpan.GetPlaces() {
+			for _, p2 := range nextSpan.GetPlaces() {
 				if containingPlace == p2.GetDcid() {
+					// If p2's place is in p1's containing places, store p1 in map.
 					dcidToRecogPlaces[p1.GetDcid()] = p1
+					// For all of p2's next containing places, add p1's DCID as the starting DCID.
+					for _, p := range p2.GetContainingPlaces() {
+						if val, ok := nextContainingPlaceToStartDcid[p]; ok {
+							nextContainingPlaceToStartDcid[p] = append(val, p1.GetDcid())
+						} else {
+							nextContainingPlaceToStartDcid[p] = []string{p1.GetDcid()}
+						}
+					}
+				}
+			}
+		}
+	}
+	return dcidToRecogPlaces, nextContainingPlaceToStartDcid
+}
+
+func combineContainedInTokens(
+	res *pb.TokenSpans_Span, 
+	spans []*pb.TokenSpans_Span, startIdx, numSpans int) *pb.TokenSpans_Span {
+	startSpan := spans[startIdx]
+	nextSpan, nextSpanIndex :=  getNextPlaceTokenSpan(spans, startIdx+1)
+
+	startingDcids := []string{}
+	for _, v := range startSpan.GetPlaces() {
+		startingDcids = append(startingDcids[:], v.GetDcid())
+	}
+
+	// This map is used to collect all the places for the combined span, with dedup.
+	dcidToRecogPlaces, nextContainingPlaceToStartDcid := findRecogPlaces(startSpan, nextSpan)
+	
+	if numSpans == 5 {
+		nextNextSpan, _ := getNextPlaceTokenSpan(spans, nextSpanIndex+1)
+		for _, actualPlace := range nextNextSpan.GetPlaces() {
+			// Iterate through all possible places in the 3rd place span.
+			for possibleNextContaining, startDcids := range nextContainingPlaceToStartDcid {
```

Comment:
since nextContainingPlaceToStartDcid is a map of nextContainingPlace to list of startDcids, can't we use actualPlace.GetDcid() to get the list of startDcids like:

```
startDcids, ok := nextContainingPlaceToStartDcid[actualPlace.getDcid()]
if !ok {
continue
}
for _, dcid := range startDcids {
...
}
```

Reply:
This comment helped a lot! This is what I meant to do when I reversed the key-val ordering in the map but then got mixed up and did it wrong. I was able to remove the two most inner loops and the equality constraint.
Should be good now, thanks!